File: LeetCode/javascript/lc_0111.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

/**
 * BFS：找到第一个叶子节点
 */
function minDepth(root: TreeNode | null): number {

    let queue = [];
    root && queue.push(root);
    let level = 0;
    while (queue.length) {
        level++;
        let size = queue.length;
        while (size--) {
            const cur = queue.shift();
            if (cur.left == null && cur.right == null) { return level; }
            if (cur.left) queue.push(cur.left);
            if (cur.right) queue.push(cur.right);
        }
    }
    return level;
};

/**
 * DFS
 */
function minDepth(root: TreeNode | null): number {
    if (root == null) return 0;
    const leftDepth = minDepth(root.left);
    const rightDepth = minDepth(root.right);
    return ((leftDepth && rightDepth) ? Math.min(leftDepth, rightDepth) : rightDepth + leftDepth) + 1;
    // return (leftDepth == 0 ? rightDepth : rightDepth == 0 ? leftDepth : Math.min(leftDepth, rightDepth)) + 1;
};

File: LeetCode/cpp/lc_0111.cpp
/**
 * @file     lc_0111.cpp
 * @brief    https://leetcode-cn.com/problems/minimum-depth-of-binary-tree/
 * @author   YongDu
 * @date     2021-09-09
 */

class Solution {
public:
  int minDepth(TreeNode *root) {
    if (root == nullptr) {
      return 0;
    }
    if (root->left != nullptr && root->right != nullptr) {
      return std::min(minDepth(root->left), minDepth(root->right)) + 1;
    }
    return std::max(minDepth(root->left), minDepth(root->right)) + 1;
  }
};

// 层序遍历版
class Solution {
public:
  int minDepth(TreeNode *root) {
    if (nullptr == root) {
      return 0;
    }
    int depth = 0;
    std::queue<TreeNode *> que;
    que.push(root);
    while (!que.empty()) {
      int size = que.size();
      depth++;
      while (size--) {
        TreeNode *node = que.front();
        que.pop();
        if (!node->left && !node->right) { // 当左右孩子皆为空，到达最小深度
          return depth;
        }
        if (node->left) {
          que.emplace(node->left);
        }
        if (node->right) {
          que.emplace(node->right);
        }
      }
    }
    return depth;
  }
};
